src/components/structures/MessagePanel.js

```diff
@@ -294,12 +295,20 @@ module.exports = React.createClass({
 
             // Wrap consecutive member events in a ListSummary
             if (isMembershipChange(mxEv)) {
+                let ts1 = mxEv.getTs();
+
+                if (this._wantsDateSeparator(prevEvent, ts1)) {
+                    let dateSeparator = <li key={ts1+'~'}><DateSeparator key={ts1+'~'} ts={ts1}/></li>;
+                    ret.push(dateSeparator);
+                }
+
                 let summarisedEvents = [mxEv];
                 i++;
                 for (;i < this.props.events.length; i++) {
                     let collapsedMxEv = this.props.events[i];
 
-                    if (!isMembershipChange(collapsedMxEv)) {
+                    if (!isMembershipChange(collapsedMxEv) ||
+                        this._wantsDateSeparator(this.props.events[i-1], collapsedMxEv.getTs())) {
```

Comment:
can't help but feel this loop would be much clearer if you set `i=i-1`: ie, drop lines 306 and 312, change line 307 to be `i+1 < this.props.events.length`, etc.